broker.go

```diff
@@ -0,0 +1,604 @@
+		host:       host,
+		gsExchange: gs,
+		tmpDir:     tmpDir,
+
+		addrTTL:    addrTTL,
+		psub:       psub,
+		topic:      pubsubTopic,
+		cancelBrkr: cancelBrkr,
+		cancelAll:  cancelAll,
+
+		peerPolicy: peerPolicy,
+		handlers:   make(map[peer.ID]*handler),
+		inEvents:   make(chan ChangeEvent, 1),
+	}
+
+	lb.unsubEvents = dt.SubscribeToEvents(lb.onEvent)
+
+	// Start watcher to read pubsub messages.
+	go lb.watch(ctx)
+	// Start distributor to send ChangeEvent messages to interested parties.
+	go lb.distributeEvents(ctx)
+
+	return lb, nil
+}
+
+// GetLatestSync returns the latest synced CID for the specified peer. If there
+// is not handler for the peer, then nil is returned.  This does not mean that
+// no data is synced with that peer, it means that the broker does not know
+// about it.  Calling Sync() first may be necessary.
+func (lb *LegBroker) GetLatestSync(peerID peer.ID) ipld.Link {
+	lb.handlersMutex.Lock()
+	hnd, ok := lb.handlers[peerID]
+	if !ok {
+		lb.handlersMutex.Unlock()
+		return nil
+	}
+	lb.handlersMutex.Unlock()
+
+	return hnd.getLatestSync()
+}
+
+// SetLatestSync sets the latest synced CID for a specified peer.  If there is
+// no handler for the peer one is created, regardless of policy.
+func (lb *LegBroker) SetLatestSync(peerID peer.ID, latestSync cid.Cid) error {
+	hnd, err := lb.getOrCreateHandler(peerID, true)
+	if err != nil {
+		return err
+	}
+
+	hnd.setLatestSync(latestSync)
+	return nil
+}
+
+// SetPolicyHandler configures LegBroker with a function to evaluate whether to
+// allow or reject messages from the specified peer.  This replaces any
+// previously configured policy handler.
+func (lb *LegBroker) SetPolicyHandler(policy PeerPolicyHandler) {
+	lb.handlersMutex.Lock()
+	defer lb.handlersMutex.Unlock()
+	lb.peerPolicy = policy
+}
+
+// Close shuts down the broker.
+func (lb *LegBroker) Close() error {
+	lb.unsubEvents()
+	lb.psub.Cancel()
+	lb.cancelBrkr()
+
+	// Dismiss any event readers.
+	lb.outEventsMutex.Lock()
+	for _, ch := range lb.outEventsChans {
+		close(ch)
+	}
+	lb.outEventsChans = nil
+	lb.outEventsMutex.Unlock()
+
+	// Dismiss any handlers waiting completion of sync.
+	lb.syncDoneMutex.Lock()
+	if len(lb.syncDoneChans) != 0 {
+		log.Warnf("Closing broker with %d syncs in progress", len(lb.syncDoneChans))
+	}
+	for _, ch := range lb.syncDoneChans {
+		close(ch)
+	}
+	lb.syncDoneChans = nil
+	lb.syncDoneMutex.Unlock()
+
+	// Shutdown datatransfer.
+	lb.cancelAll()
+	return nil
+}
+
+// OnChange creates a channel that receives change notifications, and adds that
+// channel to the list of notification channels.
+//
+// Calling the returned cancel function removed the notification channel from
+// those the receive change notifications, and closes the channel to allow any
+// waiting goroutines to stop waiting on the channel.  For this reason, channel
+// readers should check for channel closure.
+func (lb *LegBroker) OnChange() (<-chan ChangeEvent, context.CancelFunc) {
```

Comment:
If there's a single 'onchange' here, how do we incorporate non-advertising providers?

I wonder if there shouldn't be two layers of abstraction:
One object watching pubsub and auto-discovering providers to subscribe to, that can easily feed additions and removals into another watcher that emits changes across the collection of known providers. Then a user can manually add a provider in addition to using the auto discovery in a conceptually reasonable way.

Reply:
The events here do not come from watching pubsub, but rather watching events generated by the datatransfer manager.  The channel will also receive events from providers that are manually synced to the latest, as well as those auto-discovered.  A new provider can be added manually by calling `Broker.Sync()` and passing the peer ID of the provider to discover.

Reply:
If i call `Sync` with a multiaddr that's an http multiaddr, i don't see where that would get delegated to the HTTP subscriber package?

Reply:
There is no code to do that currently.

Reply:
This is done here: https://github.com/filecoin-project/go-legs/pull/45/files#diff-879d7ff817ef6271a190861e67ab0ff3a2751136cedff764b3f680565540dc2bR329-R330